k9mail-library/src/main/java/com/fsck/k9/mail/transport/SmtpTransport.java

```diff
@@ -756,6 +829,71 @@ private void saslAuthCramMD5(String username, String password) throws MessagingE
         }
     }
 
+    private void saslXoauth2(String username) throws MessagingException, IOException {
+        retryXoauthWithNewToken = true;
+        try {
+            attemptXoauth2(username);
+        } catch (NegativeSmtpReplyException negativeResponse) {
+            if (negativeResponse.getReplyCode() != SMTP_AUTHENTICATION_FAILURE_ERROR_CODE) {
```

Comment:
I'll look at this (although it'll just involve catching the NegativeSmtpReplyException and throwing something that encapsulates it).